### PR TITLE
blocks: file_source ("file_begin" stream tag)

### PR DESCRIPTION
**Pull request by rubund**
- Add a configurable "file_begin" stream tag
  which is especially useful in repeat mode.
- Added unit tests for new functionality

### CODE REVIEW

File: gr-blocks/python/blocks/qa_file_source_sink.py [hunk 1]
```diff
@@ -104,6 +105,67 @@ def test_file_source_can_seek_after_open(self):
 
         os.remove(filename)
 
+    def test_begin_tag(self):
+        src_data = range(1000)
+        expected_result = range(1000)
+
+        filename = "tmp.32f"
```
Comment (jmcorgan) [inline gr-blocks/python/blocks/qa_file_source_sink.py]: You can use Python's tempfile module and ``with`` construct here, which will create a unique filename and remove it automatically.

Comment (rubund) [inline gr-blocks/python/blocks/qa_file_source_sink.py]: Good idea. I just did it the same way as for the other tests in qa_file_source_sink.py. Do you mean I should change it for the "old" tests as well ? Would it not be outside the scope of the pull request then ?

Comment (jmcorgan) [inline gr-blocks/python/blocks/qa_file_source_sink.py]: We have lots of bad habits strewn throughout the tree, which unfortunately get copied by newcomers.  So, go ahead and do it the correct way for this new test, and for extra credit, submit a new PR with changes to the other ones.

Comment (rubund) [inline gr-blocks/python/blocks/qa_file_source_sink.py]: I have now done what you suggested. There is a new PR for the old parts of the test here: https://github.com/gnuradio/gnuradio/pull/1311

**Comment by jmcorgan** [comment]
I'd like to see the begin tag configurable instead of statically set.  This would also eliminate the need for two different bools to track config/state.  The tag id would start out as NIL; the set_begin_tag() call would take a pmt and set the key id (which would become a member variable instead of static). Finally, the work function would test if the key id was NIL to see if the feature was turned on.

This also would mean the GRC XML would need to test if the param was supplied or not and only conditionally call the set_begin_tag() method.

**Comment by rubund** [comment]
I pushed a commit where set_begin_tag has been converted to accept a pmt::pmt_t instead bool. Please let me know if I understood your suggestion correctly.

> This also would mean the GRC XML would need to test if the param was supplied or not and only conditionally call the set_begin_tag() method.

I do actually not know how to do this, nor have I been able to found this being done for any other blocks. Can you give me an example?

Currently, the GRC model is non-working if no argument is supplied.

**Pull request merged**